astropy/utils/data_info.py

```diff
@@ -0,0 +1,303 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# Note: these functions and classes are tested extensively in astropy table
+# tests via their use in providing mixin column info, and in
+# astropy/tests/test_info for providing table and column info summary data.
+
+import os
+import sys
+from copy import deepcopy
+import weakref
+import numpy as np
+from functools import partial
+
+from ..extern import six
+from ..utils import OrderedDict
+from ..utils.compat import NUMPY_LT_1_8
+
+def data_info_factory(names, funcs):
+    """
+    Factory to create a function that can be used as an ``option``
+    for outputting data object summary information.
+
+    Examples
+    --------
+    >>> from astropy.utils.data_info import data_info_factory
+    >>> from astropy.table import Column
+    >>> c = Column([4., 3., 2., 1.])
+    >>> mystats = data_info_factory(names=['min', 'median', 'max'],
+    ...                             funcs=[np.min, np.median, np.max])
+    >>> c.info(option=mystats)
+    min = 1.0
+    median = 2.5
+    max = 4.0
+    n_bad = 0
+    length = 4
+
+    Parameters
+    ----------
+    names: list
+        List of information attribute names
+    funcs: list
+        List of functions that compute the corresponding information attribute
+
+    Returns
+    -------
+    func: function
+        Function that can be used as a data info option
+    """
+    def func(dat):
+        outs = []
+        for name, func in zip(names, funcs):
+            try:
+                if isinstance(func, six.string_types):
+                    out = getattr(dat, func)()
+                else:
+                    out = func(dat)
+            except:
+                outs.append('--')
+            else:
+                outs.append(str(out))
+
+        return OrderedDict(zip(names, outs))
+    return func
+
+
+def _get_data_attribute(dat, attr=None):
+    """
+    Get a data object attribute for the ``attributes`` info summary method
+    """
+    if attr == 'class':
+        val = type(dat).__name__
+    elif attr == 'dtype':
+        val = dat.info.dtype.name
+    elif attr == 'shape':
+        datshape = dat.shape[1:]
+        val = datshape if datshape else ''
+    else:
+        val = getattr(dat.info, attr)
+    if val is None:
+        val = ''
+    return str(val)
+
+
+class InfoDescriptor(object):
```

Comment:
I left a comment on some commit somewhere questioning the need for this class, versus just making `DataInfo` itself the descriptor.  But now I can't find that original comment anymore because GitHub.

Anyways, I see now why it was done this way--it seems to me you wanted to be able to instantiate a DataInfo only when this attribute is accessed via an instance of the class that this descriptor belongs to, and then to attach the `_parent_ref`.

Instead, if `DataInfo` (and friends) are themselves descriptors, you can still have an instance of the descriptor that lives in the class definition like:
```python
class Table(object):
    info = TableInfo()
```
this `TableInfo` is referred to as an *unbound descriptor* similarly to "unbound method".  It's like a template for later descriptors that will be *bound* to individual instances of the class.  Accessing most attributes or methods of the unbound descriptor should probably fail (for example by checking if `self._parent_ref is None`).  Though the unbound descriptor could still have some use--for example its `__repr__` can display what attributes the info object carries.

The `__get__` method can then return a *bound* version of the descriptor of the same class, the only difference being that it has its `._parent_ref` set.  Incidentally the `__get__` implemented below will break when you try to look at the descriptor through the class:

```python
In [1]: from astropy.table import Table

In [2]: Table.info
ERROR: AttributeError: 'NoneType' object has no attribute '__dict__' [astropy.utils.data_info]
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-915a8e227a9a> in <module>()
----> 1 Table.info

/internal/1/root/src/astropy/astropy/astropy/utils/data_info.py in __get__(self, instance, owner_cls)
    154 
    155     def __get__(self, instance, owner_cls):
--> 156         if 'info' not in instance.__dict__:
    157             instance.__dict__['info'] = self.info_cls()
    158         instance.__dict__['info']._parent = instance

AttributeError: 'NoneType' object has no attribute '__dict__'
```

This is because accessing a descriptor through the class goes through `__get__` too, but `instance` is `None` in this case.  It's typical for this case to just:
```python
if instance is None:
    return self
```
where `self` is the unbound descriptor.

Then, to make the bound descriptor you do something like:
```
info = instance.__dict__.get('info')
if info is None:
    info = self.__class__()  # Or copy or whatever
    info._parent_ref = weakref.ref(instance)
    instance.__dict__['info'] = info

return info
```

The one downside to the above approach is that the `__init__` of the info class can't rely on `_parent_ref`.  If this is undesirable you could also use the ugly but perfectly valid approach of:
```
info = self.__new__(self.__class__)
info._parent_ref = weakref.ref(instance)
info.__init__()
```
else:
    inst = instance.__dict__['info']